geomesa-fs/geomesa-fs-storage/geomesa-fs-storage-common/src/main/scala/org/locationtech/geomesa/fs/storage/common/FileSystemThreadedReader.scala

```diff
@@ -8,101 +8,327 @@
-                  count += 1
-                }
+      try {
+        readers.foreach { case (reader, files) =>
+          // group our files by actions which can be parallelized
+          val groups = scala.collection.mutable.ListBuffer.empty[Seq[StorageFilePath]]
+          var group = scala.collection.mutable.ArrayBuffer.empty[StorageFilePath]
+          // ensure files are sorted in reverse chronological order
+          files.sorted.foreach { file =>
+            if (file.file.action == StorageFileAction.Append) {
+              group += file
+            } else {
+              if (group.nonEmpty) {
+                groups += group
+                group = scala.collection.mutable.ArrayBuffer.empty[StorageFilePath]
               }
-              logger.debug(s"File $path produced $count records")
-            } catch {
-              case NonFatal(e) => logger.error(s"Error reading file $path", e)
+              groups += Seq(file)
             }
           }
+          if (group.nonEmpty) {
+            groups += group // add the last group
+          }
+
+          es.submit(new ChainedReaderTask(es, phaser, reader, groups.head, groups.tail, queue))
         }
-        es.submit(runnable)
+      } catch {
+        case NonFatal(e) => es.shutdownNow(); throw e
+      } finally {
+        phaser.arriveAndDeregister()
       }
+
+      new FileSystemThreadedReader(es, phaser, queue)
     }
-    es.shutdown()
+  }
 
-    override def hasNext: Boolean = {
-      if (current != null) {
-        return true
+  /**
+    * Reads a file
+    *
+    * @param reader reader
+    * @param file file to read
+    * @param mods collection to track modifications
+    * @return
+    */
+  private def read(
+      reader: FileSystemPathReader,
+      file: StorageFilePath,
+      mods: scala.collection.mutable.Set[String]): CloseableIterator[SimpleFeature] = {
+    file.file.action match {
+      case StorageFileAction.Append => new AppendingReaderIterator(reader, file.path, mods)
+      case StorageFileAction.Modify => new ModifyingReaderIterator(reader, file.path, mods)
+      case StorageFileAction.Delete => new DeletingReaderIterator(reader, file.path, mods)
+      case _ => throw new NotImplementedError(s"Unexpected storage action: ${file.file.action}")
+    }
+  }
+
+  /**
+    * Performs a set of chained (dependent) reads. The groups must be handled serially, but the files within a
+    * given group can be read in parallel.
+    *
+    * @param es executor service for submitting new read tasks
+    * @param phaser phaser to track run lifecycle
+    * @param reader reader
+    * @param group current group of files that can be read in parallel
+    * @param chain remaining groups of files that must be read sequentially
+    * @param queue result queue
+    * @param mods modifications/deletes for this group of files
+    */
+  private class ChainedReaderTask(
+      es: ExecutorService,
+      phaser: Phaser,
```

Comment:
Is the Phaser required?  Or could the same thing be achieved by reading all the files in this 'group' and then going from there?

Reply:
That is basically what is being done. This phaser tracks the overall completion of all groups - it's the same one that shuts down the executor service, above. Each group of files will register with the overall phaser, read the group in parallel, kick off the next sequential group (which will register itself on creation), then de-register itself. In that way, when the phaser is finally triggered, we know that all groups have completed.